### PR TITLE
FI-435 Track assertions in helpers

### PR DESCRIPTION
**Pull request by Jammjammjamm**
This PR updates the assertion tracking to track assertion calls in helper methods. For assertions not directly called in a test, the output will now look like this:
```
location_of_method_call_in_test -> location_of_assertion
modules/argonaut/argonaut_procedure_sequence.rb:110 => sequence_base.rb:490
```
This allows us to verify that the helper methods are being exercised in each test, rather than only once. The output also differentiates between direct and indirect assertion calls.

A unit test of the following test would record a direct assertion call because the item being unit tested is making the assertion.
```
test 'direct' do
  assert true
end
```

A unit test of the following test would record an indirect assertion call because the assertion happens in a separate method. A unit test of the `validate` method would record the assertion as a direct assertion because the assertion is called within that method.
```
test 'indirect' do
  validate
end

def validate
  assert true
end
```
Differentiating direct and indirect assertions allows us to determine whether helper methods are themselves being unit tested or only being tested through the tests that use them.

**Submitter:**
- [x] This pull request describes why these changes were made
- [x] Internal ticket for this PR: https://oncprojectracking.healthit.gov/support/browse/FI-435
- [x] Internal ticket links to this PR
- [x] Internal ticket is properly labeled (Community/Program)
- [x] Internal ticket has a justification for its Community/Program label
- [x] Code diff has been reviewed for extraneous/missing code
- N/A Tests are included and test edge cases
- [x] Tests/code quality metrics have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

### CODE REVIEW

**Comment by Jammjammjamm** [comment]
I think capturing the assertion messages is something that would require more discussion. Many of the assertion messages only make sense when the assertion fails. Updating them to be meaningful for passes and failures would require a lot of effort and probably be difficult to do without making them less understandable to end users.

Many of the assertion messages are dynamic, so we'd need to capture them all individually and determine the best way to format the output.

**Comment by radamson** [comment]
I just tried running this and when I looked at the output the first entry I saw was:

`modules/argonaut/argonaut_allergy_intolerance_sequence.rb:107 => sequence_base.rb:514` with 1 Direct Pass Count and, 0 Direct Fail Counts and no Indirect results.

I thought that this would be an example of an indirect count looking at the code.  Am I misinterpreting something here?

**Comment by Jammjammjamm** [comment]
The place where the direct/indirect distinction shows up isn't in the individual results for an assertion that goes through a helper function, such as
```
modules/smart/token_refresh_sequence.rb:85 => modules/smart/token_refresh_sequence.rb:113,11,2
```
Rather, the indirect results for `modules/smart/token_refresh_sequence.rb:113`:
```
modules/smart/token_refresh_sequence.rb:113,12,1,22,4
```
are just the sum of the results from all the indirect calls to it:
```
modules/smart/token_refresh_sequence.rb:109 => modules/smart/token_refresh_sequence.rb:113,11,2
modules/smart/token_refresh_sequence.rb:85 => modules/smart/token_refresh_sequence.rb:113,11,2
```

File: test/support/sequence_coverage_reporting.rb [hunk 1]
```diff
@@ -60,49 +65,141 @@ def add_assertion_location(location)
...
-    def pass_count(results)
-      results.count { |result| result }
+    def direct_call?(location)
+      location.include? AssertionCallLocationFormatter::CALL_SEPARATOR
     end
 
-    def fail_count(results)
-      results.count(&:!)
+    def direct_pass_count(location)
+      AssertionTracker.assertion_calls[location].count { |result| result }
```
Comment (radamson) [inline test/support/sequence_coverage_reporting.rb]: Could also use `.count(&:itself)`.  Not necessarily suggesting it, but it might fit well with how you are counting fails `.count(&:!)` already.

File: test/support/sequence_coverage_reporting.rb [hunk 2]
```diff
@@ -60,49 +65,141 @@ def add_assertion_location(location)
...
+      AssertionTracker.assertion_calls.reduce(0) do |count, (key, results)|
+        key.end_with?(location) && key != location ? count + results.count { |result| result } : count
+      end
+    end
+
+    def indirect_fail_count(location)
+      return if direct_call? location
+
+      AssertionTracker.assertion_calls.reduce(0) do |count, (key, results)|
+        key.end_with?(location) && key != location ? count + results.count(&:!) : count
```
Comment (radamson) [inline test/support/sequence_coverage_reporting.rb]: This line can be a little tricky to read with all the operations going on. I don't think it has to change, but even moving `key.end_with?(location) && key != location` out might make it more clear, although at the expense of another line.

File: test/support/sequence_coverage_reporting.rb [hunk 3]
```diff
@@ -60,49 +65,141 @@ def add_assertion_location(location)
...
+      self.sequence_call_index = index if backtrace[index].to_s.match?(/block .*in #{method_name}/)
+    end
+  end
+
+  def location
+    return unless sequence_call_index.present? && assertion_call_index.present?
+
+    [backtrace[sequence_call_index], backtrace[assertion_call_index + 1]]
+      .uniq
+      .map { |location| location&.to_s&.match(LINE_REGEX)&.[](1) }
```
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply test/support/sequence_coverage_reporting.rb]: I actually hate it, but it gets the job done.

File: test/support/sequence_coverage_reporting.rb (hunk 1 above)
Comment (Jammjammjamm) [reply test/support/sequence_coverage_reporting.rb]: Nice, didn't know about that one.

File: test/support/sequence_coverage_reporting.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (radamson) [reply test/support/sequence_coverage_reporting.rb]: You could also use `#dig` here, which [Matz himself _might_ suggest](https://bugs.ruby-lang.org/issues/11813).

[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply test/support/sequence_coverage_reporting.rb]: Yeah, I think dig would look nicer, so I'll use that.

[the remark this answers is not in the document]
Comment (radamson) [reply test/support/sequence_coverage_reporting.rb]: the final way you selected with `captures` is even more readable 👍
